chrome/content/zotero/xpcom/itemTreeManager.js

```diff
@@ -0,0 +1,379 @@
+	 *     minWidth: 50, // minimum width in pixels
+	 *     // iconPath: 'chrome://zotero/skin/tick.png', // icon to show in the column header
+	 *     htmlLabel: '<span style="color: red;">reversed title</span>', // use HTML in the label. This will override the label and iconPath property
+	 *     ignoreInColumnPicker: false, // show in the column picker
+	 *     submenu: true, // show in the column picker submenu
+	 *     primary: false, // only one primary column is allowed
+	 *     pluginID: 'make-it-red@zotero.org', // plugin ID, which will be used to unregister the column when the plugin is unloaded
+	 *     dataProvider: (item, dataKey) => {
+	 *         // item: the current item in the row
+	 *         // dataKey: the dataKey of the column
+	 *         // return: the data to display in the column
+	 *         return item.getField('title').split('').reverse().join('');
+	 *     },
+	 *     zoteroPersist: ['width', 'hidden', 'sortDirection'], // persist the column properties
+	 * });
+	 * ```
+	 * @example
+	 * Register multiple custom columns:
+	 * ```js
+	 * const registeredDataKeys = await Zotero.ItemTreeManager.registerColumns(
+	 * [
+	 *     {
+	 *          dataKey: 'rtitle',
+	 *          iconPath: 'chrome://zotero/skin/tick.png',
+	 *          label: 'reversed title',
+	 *          pluginID: 'make-it-red@zotero.org', // Replace with your plugin ID
+	 *          dataProvider: (item, dataKey) => {
+	 *              return item.getField('title').split('').reverse().join('');
+	 *          },
+	 *     },
+	 *     {
+	 *          dataKey: 'utitle',
+	 *          label: 'uppercase title',
+	 *          pluginID: 'make-it-red@zotero.org', // Replace with your plugin ID
+	 *          dataProvider: (item, dataKey) => {
+	 *              return item.getField('title').toUpperCase();
+	 *          },
+	 *     },
+	 * ]);
+	 * ```
+	 */
+	async registerColumns(options) {
+		const registeredDataKeys = this._addColumns(options);
+		if (!registeredDataKeys) {
+			return false;
+		}
+		this._addPluginShutdownObserver();
+		await this._notifyItemTrees();
+		return registeredDataKeys;
+	}
+
+	/**
+	 * Unregister a custom column.
+	 * Although it's async, resolving does not promise the item trees are updated.
+	 * @param {string | string[]} dataKeys - The dataKey of the column to unregister
+	 * @returns {boolean} true if the column(s) are unregistered
+	 * @example
+	 * ```js
+	 * Zotero.ItemTreeManager.unregisterColumns('rtitle');
+	 * ```
+	 */
+	async unregisterColumns(dataKeys) {
+		const success = this._removeColumns(dataKeys);
+		if (!success) {
+			return false;
+		}
+		await this._notifyItemTrees();
+		return true;
+	}
+
+	// TODO: add cell renderer registration
+
+	/**
+	 * Get column(s) that matches the properties of option
+	 * @param {undefined | Partial.<ItemTreeColumnOptions> | Partial.<ItemTreeColumnOptions>[]} options - An option or array of options to match
+	 * @returns {ItemTreeColumnOptions[]}
+	 */
+	getCustomColumns(options) {
```

Comment:
I have to stress again that this is over-engineered based on how it will primarily be used. `itemTreeID` should be the primary filter function here as the first, separate argument. Then the results can be additionally filtered on `options`, although I am happy to just drop that completely. Also the function should not take arrays as arguments.

Reply:
I use arguments object (`options`) instead of arguments list here because:
1. All filters, i.e. enabledTreeID, dataKey, pluginID, ... are optional
2. The arguments may grow in the future
3. arguments object makes the code more readable. Consider this: `getCustomColumns('my-title', 'show-title')` v.s. `getCustomColumns({dataKey: 'my-title', enabledTreeIDs: 'show-title'})`

If we rewrite this method with each filter key we want to use, e.g. enabledTreeID, dataKey, pluginID, the implementation would be more lengthy because we will need to write filter functions (filter by dataKey, by enabledTreeIDs, by ...) for each filter key, not just in a for-loop of all arguments in the arguments object.

What do you think?

Reply:
> 1. All filters, i.e. enabledTreeID, dataKey, pluginID, ... are optional

Yes, but treeID should be a separate, first argument, and the required filter



> 2\. The arguments may grow in the future

I don't see this happening. We do not expect to cover every possible way plugins may augment Zotero with this API, only the generic case for the majority of plugins. As such, this function is expected to only really be useful to the ItemTree when getting a full list of columns to display and thus only requires the ability to filter by `treeID`.



> 3\. arguments object makes the code more readable. Consider this: `getCustomColumns('my-title', 'show-title')` v.s. `getCustomColumns({dataKey: 'my-title', enabledTreeIDs: 'show-title'})`

Right, but I proposed an API like `getCustomColumns('item-tree-id', {additionalFilterDataKey: "key", additionalFilter2: "filter"})`, although I don't think we need to even add a way to filter additionally at all here, since our code doesn't need it and the majority of plugins won't either, although if they do, they can filter in their own code just as good.

Reply:
Done